### PR TITLE
scanoss: Replace usages of raw JSON by model classes for ScanResponse

### PR DESCRIPTION
**Pull request by nnobelis**
Please refer to individual commit messages.

### CODE REVIEW

File: clients/scanoss/src/main/kotlin/model/Model.kt [hunk 1]
```diff
@@ -67,6 +67,17 @@ data class License(
     val osadlUpdated: String? = null
 )
 
+/**
```
Comment (sschuberth) [inline clients/scanoss/src/main/kotlin/model/Model.kt:70]: Commit message nits:
- "of *the* scanoss-junit-4.12.json` test file"
- "by *the* SCANOSS team"
- "in a new commit" is clear and could be dropped IMO

File: clients/scanoss/src/main/kotlin/ScanOssService.kt [hunk 2]
```diff
@@ -34,6 +34,8 @@ import retrofit2.http.Multipart
 import retrofit2.http.POST
 import retrofit2.http.Part
 
+typealias FullScanResponse = Map<String, List<ScanResponse>>
```
Comment (sschuberth) [inline clients/scanoss/src/main/kotlin/ScanOssService.kt:37]: Please move the introduction of this `typealias` to a separate preceding commit, so that this commit really only touches files in the "scanner" directory.

File: scanner/build.gradle.kts [hunk 3]
```diff
@@ -85,3 +90,11 @@ buildConfig {
     buildConfigField("String", "SCANCODE_VERSION", "\"$scancodeVersion\"")
     buildConfigField("String", "SCANOSS_VERSION", "\"$scanossVersion\"")
 }
+
+tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
```
Comment (sschuberth) [inline scanner/build.gradle.kts:94]: Please use an import instead of a full qualified name (also see #5234).

File: scanner/build.gradle.kts [hunk 4]
```diff
@@ -85,3 +90,11 @@ buildConfig {
...
+    if (name.contains("Test")) {
```
Comment (sschuberth) [inline scanner/build.gradle.kts:95]: I believe it's fine to unconditionally do this for all tasks. That would also align with existing similar code.

Comment (sschuberth) [reply scanner/build.gradle.kts:95]: Speaking of aligning code, please also use the `customCompilerArgs` approach here.

File: scanner/src/main/kotlin/scanners/scanoss/ScanOss.kt [hunk 5]
```diff
@@ -120,16 +115,11 @@ class ScanOss internal constructor(
                 "The $scannerName server returned an UUID not present in the mapping."
             }
 
-            fileNamesAnonymizationMapping[uuid] to entry.value
+            fileNamesAnonymizationMapping[uuid]!! to entry.value
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/scanoss/ScanOss.kt:118]: Instead of introducing `!!`, how about (in a preceeding commit) replacing the code incl. `require()` above with

            val fileName = fileNamesAnonymizationMapping[uuid] ?: throw IllegalArgumentException(
                "The $scannerName server returned an UUID not present in the mapping."
            )

            fileName to entry.value

File: scanner/src/test/assets/scanoss-junit-4.12.json [hunk 6]
```diff
@@ -2,6 +2,22 @@
   "/tmp/ort-ScanOss2759786101559527642/Maven/junit/junit/4.12/to-do.txt": [
     {
       "id": "none",
+      "status": "pending",
```
Comment (sschuberth) [inline scanner/src/test/assets/scanoss-junit-4.12.json:5]: Why was this file modified? Where do these changes come from? Maybe do this in a separate preceding commit with more explanations in the commit message?

File: scanner/src/test/kotlin/scanners/scanoss/ScanOssResultParserTest.kt [hunk 7]
```diff
@@ -28,17 +28,23 @@ import io.kotest.matchers.should
...
+import kotlinx.serialization.json.decodeFromStream
+
+import org.ossreviewtoolkit.clients.scanoss.FullScanResponse
 import org.ossreviewtoolkit.model.CopyrightFinding
 import org.ossreviewtoolkit.model.LicenseFinding
 import org.ossreviewtoolkit.model.TextLocation
-import org.ossreviewtoolkit.model.readJsonFile
 import org.ossreviewtoolkit.utils.spdx.SpdxConstants
 
+private val json = Json { ignoreUnknownKeys = true }
```
Comment (sschuberth) [inline scanner/src/test/kotlin/scanners/scanoss/ScanOssResultParserTest.kt:40]: Please use `ScanOssService.JSON` instead.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/build.gradle.kts (hunk 4 above)
Comment (nnobelis) [reply scanner/build.gradle.kts:95]: > I believe it's fine to unconditionally do this for all tasks. That would also align with existing similar code.

It does not work if I remove the conditional: the `kotlinx` dependency being a test dependency (no release runtime code uses it), it tries to build non-test classes with the `optIn` and fails.

Comment (sschuberth) [reply scanner/build.gradle.kts:101]: Interesting. Then please at lest use `"test" in name.lowerCase()` instead.

Comment (nnobelis) [reply scanner/build.gradle.kts]: Done but Intellij does not recognize `lowerCase`: I am forced to use the Java version. Any idea ?

Comment (sschuberth) [reply scanner/build.gradle.kts]: Ah, might be that for Gradle KTS files we're stuck with a Kotlin version where it's still called `toLowerCase()`.

Comment (nnobelis) [reply scanner/build.gradle.kts]: Can we merge this ?

**Review by sschuberth — APPROVED** [review]
